source/server/active_stream_socket.h

```diff
@@ -0,0 +1,222 @@
+  public:
+    GenericListenerFilter(const Network::ListenerFilterMatcherSharedPtr& matcher,
+                          Network::ListenerFilterPtr listener_filter)
+        : listener_filter_(std::move(listener_filter)), matcher_(std::move(matcher)) {}
+    Network::FilterStatus onAccept(ListenerFilterCallbacks& cb) override {
+      if (isDisabled(cb)) {
+        return Network::FilterStatus::Continue;
+      }
+      return listener_filter_->onAccept(cb);
+    }
+    /**
+     * Check if this filter filter should be disabled on the incoming socket.
+     * @param cb the callbacks the filter instance can use to communicate with the filter chain.
+     **/
+    bool isDisabled(ListenerFilterCallbacks& cb) {
+      if (matcher_ == nullptr) {
+        return false;
+      } else {
+        return matcher_->matches(cb);
+      }
+    }
+
+  private:
+    const Network::ListenerFilterPtr listener_filter_;
+    const Network::ListenerFilterMatcherSharedPtr matcher_;
+  };
+  using ListenerFilterWrapperPtr = std::unique_ptr<GenericListenerFilter>;
+
+  // Network::ListenerFilterManager
+  void addAcceptFilter(const Network::ListenerFilterMatcherSharedPtr& listener_filter_matcher,
+                       Network::ListenerFilterPtr&& filter) override {
+    accept_filters_.emplace_back(
+        std::make_unique<GenericListenerFilter>(listener_filter_matcher, std::move(filter)));
+  }
+
+  // Network::ListenerFilterCallbacks
+  Network::ConnectionSocket& socket() override { return *socket_.get(); }
+  Event::Dispatcher& dispatcher() override;
+  void continueFilterChain(bool success) override;
+  void setDynamicMetadata(const std::string& name, const ProtobufWkt::Struct& value) override;
+  envoy::config::core::v3::Metadata& dynamicMetadata() override {
+    return stream_info_->dynamicMetadata();
+  };
+  const envoy::config::core::v3::Metadata& dynamicMetadata() const override {
+    return stream_info_->dynamicMetadata();
+  };
+
+  StreamInfo::FilterState& filterState() override { return *stream_info_->filterState().get(); }
+
+  ActiveStreamListenerBase& listener_;
+  Network::ConnectionSocketPtr socket_;
+  const bool hand_off_restored_destination_connections_;
+  std::list<ListenerFilterWrapperPtr> accept_filters_;
+  std::list<ListenerFilterWrapperPtr>::iterator iter_;
+  Event::TimerPtr timer_;
+  std::unique_ptr<StreamInfo::StreamInfo> stream_info_;
+  bool connected_{false};
+};
+
+using ActiveInternalSocket = ActiveStreamSocket;
+
+class ActiveStreamListenerBase : public ActiveListenerImplBase {
+public:
+  ActiveStreamListenerBase(Network::ConnectionHandler& parent, Event::Dispatcher& dispatcher,
+                           Network::ListenerPtr&& listener, Network::ListenerConfig& config);
+  static void emitLogs(Network::ListenerConfig& config, StreamInfo::StreamInfo& stream_info);
+
+  virtual void incNumConnections() PURE;
+  virtual void decNumConnections() PURE;
+  virtual Event::Dispatcher& dispatcher() PURE;
```

Comment:
it seems odd that ActiveStreamListenerBase has a dispatcher_ member, but dispatcher() is a PURE function.

I would suggest implementing the 3 functions above in this class instead of requiring all subclasses to provide redundant implementations.